### PR TITLE
api: unify error handling

### PR DESCRIPTION
**Pull request by kchibisov**
Make error infrastructure more backend agnostic and let backends just forward the os errors opaquely.

--

Only done for X11/Wayland for now.

### CODE REVIEW

File: src/error.rs [hunk 1]
```diff
@@ -1,132 +1,125 @@
...
-        OsError { line, file, error }
-    }
+/// A general error that may occur during a request to the windowing system.
+#[derive(Debug)]
+#[non_exhaustive]
+pub enum RequestError {
+    /// The request is not supported.
+    NotSupported(NotSupportedError),
+    /// The request was ignored by the operating system.
+    Ignored,
```
Comment (madsmtm) [inline src/error.rs:51]: I do not think this is something that we should error on? If it's ignored, it's just ignored.

**Review by madsmtm — COMMENTED** [review]
Instead of trying to merge the error structs, I think I'd rather go in the direction of splitting them up more. But it's a difference in approach, I'll try to outline the benefits of each here:

Many error types for each case:
- Easier for user to know exactly which kind of error happened, doesn't have to handle unknown cases.
- Similar to above: Easier for user to know which kinds of errors _cannot_ happen.

Roughly one for the whole crate:
- Easier to maintain.
- Easier to make changes to error handling.
- Easier for user to add conversion to their own error type (only ~one `From<WinitError>` impl needed).

Are there benefits / downsides here that I've missed? Just so that we have the same starting point for this discussion.

Also, I _feel_ like we've had this kind of discussion before, linking at least https://github.com/rust-windowing/winit/pull/2646 and https://github.com/rust-windowing/winit/pull/3481.

---

I think I'm imagining roughly the following error types:
- `EventLoopCreationError`
- `EventLoopRunError`
- `WindowError` (used in `create_window` and , similar to `RequestError` in this PR)
- `MonitorError`
- `CursorError` (`create_custom_cursor`).
- `BadImage` (`CustomCursor::from_rgba`).

But perhaps these would basically internally all be the same error? Is there something I'm missing here?

**Comment by kchibisov** [comment]
I'm fine with adding more errors, it's just `RequestError` is used for all the requests and creating a `Window` is a _request_. While I could add `Creation` error, the thing is that they likely just use os error...

We could return Os error directly, but I've used request because of `non_exhaustive`.

The point with this PR though, is not to change how errors work, but rather remove the dependency on the `platform::OsError` and a like, so our top-level errors could be used by the backends directly.

I can change things in the future to whatever errors we prefer, but for now it's more like a one to one transition + removal of `ExternalError` in favor of `RequestError` since `ExternalError` doesn't mean much as a name.

File: src/error.rs (hunk 1 above)
Comment (kchibisov) [reply src/error.rs:51]: I just moved what was used already.

**Comment by madsmtm** [comment]
> The point with this PR though, is not to change how errors work

Fair enough, apologies for the confusion.

**Comment by kchibisov** [comment]
Implemented all the backends.

**Review by MarijnS95 — APPROVED** [review]
Approving for android, some general feedback on `src/error.rs` changes.

File: src/error.rs [hunk 2]
```diff
@@ -1,132 +1,130 @@
...
-pub struct OsError {
-    line: u32,
-    file: &'static str,
-    error: platform_impl::OsError,
-}
-
-/// A general error that may occur while running the Winit event loop
+/// A general error that may occur while running or creating
+/// the event loop.
 #[derive(Debug)]
```
[the remark this answers is not in the document]
Comment (kchibisov) [reply src/error.rs:6]: Not clear what you want here, `Debug` is required to implement `Error` trait.

File: src/platform_impl/android/mod.rs [hunk 3]
```diff
@@ -14,7 +14,7 @@ use tracing::{debug, trace, warn};
 use crate::application::ApplicationHandler;
 use crate::cursor::Cursor;
 use crate::dpi::{PhysicalPosition, PhysicalSize, Position, Size};
-use crate::error::{self, EventLoopError, ExternalError, NotSupportedError};
+use crate::error::{self, EventLoopError, NotSupportedError, RequestError};
 use crate::event::{self, Force, InnerSizeWriter, StartCause};
```
[the remark this answers is not in the document]
Comment (kchibisov) [reply src/platform_impl/android/mod.rs]: No, removed that as well.

File: src/error.rs [hunk 4]
```diff
@@ -1,132 +1,130 @@
...
+        match self {
+            EventLoopError::RecreationAttempt => write!(f, "EventLoop can't be recreated"),
+            EventLoopError::Os(err) => err.fmt(f),
+            EventLoopError::ExitFailure(status) => write!(f, "Exit Failure: {status}"),
+            EventLoopError::NotSupported(err) => err.fmt(f),
+        }
+    }
+}
+
+impl Error for EventLoopError {}
```
Comment (MarijnS95) [inline src/error.rs:30]: Shouldn't we implement `fn source()` here?

File: src/error.rs [hunk 5]
```diff
@@ -1,132 +1,130 @@
...
+impl Display for RequestError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            RequestError::NotSupported(err) => err.fmt(f),
+            RequestError::Ignored => write!(f, "The request was ignored"),
+            RequestError::Os(err) => err.fmt(f),
+        }
+    }
 }
+impl Error for RequestError {}
```
Comment (MarijnS95) [inline src/error.rs:65]: And `fn source()` here too for `::Os`.

File: src/error.rs [hunk 6]
```diff
@@ -1,132 +1,130 @@
...
 
-#[allow(unused_macros)]
-macro_rules! os_error {
-    ($error:expr) => {{
-        crate::error::OsError::new(line!(), file!(), $error)
-    }};
+impl Display for RequestError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            RequestError::NotSupported(err) => err.fmt(f),
```
Comment (MarijnS95) [inline src/error.rs:59]: No complaints from `clippy::use_self`?

File: src/error.rs [hunk 7]
```diff
@@ -1,132 +1,130 @@
...
-            "{:?}, {}",
-            ExternalError::NotSupported(NotSupportedError::new()),
-            ExternalError::NotSupported(NotSupportedError::new())
-        );
+impl Display for OsError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.pad(&format!("os error at {}:{}: {}", self.file, self.line, self.error))
     }
 }
+impl Error for OsError {}
```
Comment (MarijnS95) [inline src/error.rs:123]: `fn source()` for `Some(self.error)`?

File: src/error.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (MarijnS95) [reply src/error.rs:6]: What I want is a "prettier" format for `Debug` that doesn't recursively `Debug`-print the wrapped errors, as that is what an implementation that walks the `source()` tree is supposed to do.
(`fn main()` does not, though: https://doc.rust-lang.org/src/std/process.rs.html#2454)

And here's what `anyhow` does: https://docs.rs/anyhow/latest/anyhow/struct.Error.html#display-representations

But then we don't implement `fn source()` anywhere where other `Error`s are wrapped.

---

In other words, I don't think there's anything that you can do that is correct. Either we don't see the entire error for users who `Debug`-print, or we might get duplicates if the backend implementation bubbles up `anyhow` errors, or the caller wraps `anyhow` errors (whicha `anyhow` will recurse down).

File: src/error.rs (hunk 6 above)
Comment (kchibisov) [reply src/error.rs:59]: As you can see on CI - no.

Comment (MarijnS95) [reply src/error.rs:59]: (This was a silent request to `Self::` these `match` arms 😉)

**Review by notgull — APPROVED** [review]
I don't mind the heap allocation since errors are on the cold path.
